### DISCUSSION TITLE
React tsParticles Component

### DISCUSSION THREAD

**Discussion by matteobruni** [matteobruni/tsparticles#1125] [General]
This section will be used to talk about the `React` version of tsParticles, the `react-tsparticles` package.

**Comment by SupaDupa24** [comment]
quick question has anyone been able to get snow falling with either tsparticles or react particles?

**Reply by matteobruni** [reply]
You can see a sample here: https://codesandbox.io/s/condescending-brook-rr8u2?file=/src/App.js

Let me know if you need more help 😉

**Reply by bayazid18** [reply]
please give some example about bubble effect tsparticles or react particles? I'm using it in the Hero section. Position relative is not working when I am trying to absolute #tsparticles

**Reply by matteobruni** [reply]
Hi @bayazid18, as you can see [here](https://github.com/matteobruni/tsparticles/pull/2397) the `fullScreen` options have `enable: true` by default since `1.37.0`. You have to disable it if you want to customize the CSS. Another change made for that is setting by default `interactivity.detectsOn` to `"window"` by default, for your configuration it's better to set `"canvas"`. It should work as expected after these changes.

**Reply by bayazid18** [reply]
Thanks it's work fine. Where I put and test various presets like this https://tsparticles--pr2397-v1-fvk6je1n.web.app/  
I mean I want know the name of different particles effect and 2nd question does it this dependency require for various effect "tsparticles-shape-confetti" ?

**Reply by matteobruni** [reply]
> Thanks it's work fine. Where I put and test various presets like this https://tsparticles--pr2397-v1-fvk6je1n.web.app/
> I mean I want know the name of different particles effect and 2nd question does it this dependency require for various effect "tsparticles-shape-confetti" ?

Starting from version `1.30` `tsparticles-shape-confetti` was deprecated (https://dev.to/matteobruni/tsparticles-1-30-0-is-here-confetti-rework-all-particles-shapes-supported-emojis-too-2a4a).

About the samples, it’s just a drop down and some json files that I named manually, if you want some custom presets have a look here: https://particles.js.org/samples/presets/index.html

Those are packages available on npm.

**Reply by bayazid18** [reply]
Does this https://particles.js.org/samples/presets/index.html presets code sample available?

**Reply by matteobruni** [reply]
You can see some instructions here: https://github.com/matteobruni/tsparticles/tree/main/presets/bigCircles#reactjs--preact--inferno

It’s valid for all presets, just change the package and the function name

**Comment by Woverlin** [comment]
Someone please help me to create react component for this animation
https://codepen.io/matteobruni/pen/JjEddbQ
and this 
https://particles.js.org/samples/index.html#confetti
I can use snow falling above, so I think I have right config

**Reply by matteobruni** [reply]
Hi @Woverlin,

I wrote an article about that [here](https://dev.to/matteobruni/how-to-create-beautiful-confetti-animations-with-tsparticles-193)

The sample used in the article for React is [this](https://codesandbox.io/s/tsparticles-confetti-2zql1)

**Reply by Woverlin** [reply]
Thanks a lot, so I missed   `loadConfettiShape(main);` in init function,Can you share how to handle onClick in Particles react component, I use `onClick` but not work

**Reply by matteobruni** [reply]
I need to check if the functionality is enabled in v1, if not I’ll take it from the upcoming v2 and I bring it to v1

**Reply by matteobruni** [reply]
Hi @Woverlin,

I've updated the sample with the click handler, is a little tricky but it works, I'm going to make it easier to understand in a future release

**Comment by brightzoe** [comment]
hello @matteobruni ,I love your tsparticles ,it's amazing, many coolest animations drawed by it. I saw them on codepen.
I want draw rising lines use react-tsparticles,but the demo almost shape is 'circle' ,I cant draw lines by it.
The effect I want is like this :
![rise-lines](https://i.loli.net/2021/08/26/Ky4zrU5mSXDFqHg.gif)

But I‘m not familiar with it ,and have little known with canvas ,I have tried many times but no results, could I have your help , I will be greatly appreciated.
Modify on your animation,I got this ,but convert the options to react-tsparticles , there are no lines ...
https://codepen.io/brightzoe/pen/ExXavmd
The docs is hard for me , so as to ask for help.

there is a  not good practise by making many divs to fulfilled my needs:  
https://codesandbox.io/s/rise-6j3g2?file=/src/App.js
 
but how to draw by react-tsparticles?😶😖

**Reply by matteobruni** [reply]
Hi @brightzoe,

Your config needs little changes. Here's the fixed one:

``` javascript
//tsParticles library - https://github.com/matteobruni/tsparticles

tsParticles.load("tsparticles", {
  fullScreen: {
    enable: true
  },
  detectRetina: true,
  background: {
    color: "#000"
  },
  fpsLimit: 60,
  emitters: {
    direction: "top",
    life: {
      count: 0,
      duration: 0.08,
      delay: 0.1
    },
    rate: {
      delay: 0.15,
      quantity: 1
    },
    size: {
      width: 100,
      height: 0
    },
    position: {
      y: 100,
      x: 50
    }
  },
  particles: {
    number: {
      value: 0
    },
    life: {
      count: 1
    },
    shape: {
      type: "line"
    },
    size: {
      value: 50,
      animation: {
        enable: true,
        sync: true,
        speed: 1,
        startValue: "max",
        destroy: "min"
      }
    },
    stroke: {
      color: {
        value: "#ffffff"
      },
      width: 1
    },
    rotate: {
      path: true,
      value: 180
    },
    move: {
      straight: true,
      enable: true,
      speed: { min: 10, max: 20 },
      outModes: {
        default: "destroy",
        top: "none"
      },
      trail: {
        fillColor: "#000",
        enable: true,
        length: 3
      }
    }
  }
});
```

**Reply by brightzoe** [reply]
thanks very very much! these animations are magical! Your help is very quicy and helpful!

**Reply by brightzoe** [reply]
hello @matteobruni ,
sorry I'm back again. I used the config in pure React,it's ready.  https://codesandbox.io/s/rise-6j3g2?file=/src/App.js
but to my product ,I can't find why it's not working, what effects the lines?
like this,it's could't working,
https://codesandbox.io/s/elated-sid-n4vys?file=/src/App.js

circleOptions in the project good runing,but the lines is not.

**Reply by matteobruni** [reply]
I think the issue is `craco` but I don't know what it is or how it works.
If it's a SSR library this library won't work because it needs to be run client-side.
Maybe it's worth asking to the craco authors what could be the issue, since in a standard CRA application works fine.

**Reply by matteobruni** [reply]
> hello @matteobruni ,
> sorry I'm back again. I used the config in pure React,it's ready. https://codesandbox.io/s/rise-6j3g2?file=/src/App.js
> but to my product ,I can't find why it's not working, what effects the lines?
> like this,it's could't working,
> https://codesandbox.io/s/elated-sid-n4vys?file=/src/App.js
> 
> circleOptions in the project good runing,but the lines is not.

Ok I found the issue, in `1.34.0` I fixed an issue with `size` and `opacity` animation with some default values that were wrong. You have specified a `size.value` to `50`, but you have to specify a `{ min: 1, max: 50 }` range or whatever you want.
Using the range the animation from `min` to `max` will work fine.

https://codesandbox.io/s/modest-moser-kcz1o?file=/src/App.js

Here's a working sample, I changed some options to figure out what the problem was.

**Reply by brightzoe** [reply]
Thanks a lot ,it's working to change the specific value to a range, though I dont't know the inner working.🌹

`craco` is to override some configs, customization without using 'eject' by adding a single `craco.config.js` file and customize your eslint, babel, postcss configurations and many more.

Thanks for the friendly and patient guidance again.

**Comment by apudiu** [comment]
Someone please help me create [fireworks](https://particles.js.org/samples/presets/fireworks) with react component. Unfortunately copying options from [Codepen](https://codepen.io/matteobruni/pen/yLbYWbO) doesn't works :(

**Reply by brightzoe** [reply]
Hi @apudiu, Maybe like my questions above, copy options from codepen and  change `size.value` to a range. 
Then it works! 
Hope this helps. (*^_^*)

https://codesandbox.io/s/rise-6j3g2?file=/src/App.js

**Reply by apudiu** [reply]
Thank you very much. That worked, but is that mentioned anywhere in the docs?

**Comment by romjanxr** [comment]
how can I use on a single component instead of whole page?

**Reply by matteobruni** [reply]
Hi @romjanxr,

There's a `fullScreen` option to disable for having particles in a classic div. An user asked about it [here](https://github.com/matteobruni/tsparticles/discussions/2591#discussioncomment-1698267) hope it helps

**Comment by Eddie-ui** [comment]
imported a preset as json and added it to the particles component everything works fine, but when i refresh my project the particles animation stops working, i have to change a value in the particles params before it starts working again

**Reply by matteobruni** [reply]
Do you have some code to see how to reproduce the issue?

**Comment by norvalbv** [comment]
Hi,

So I'm trying to make the particles the background in the intro/hero.

I have read above that I can remove the fullscreen by changing the property. However, when I do so a few things happen. The particles effect seems to be parallax on fullscreen, not sure if this is true and if it's possible to make this on a fixed size with editing some properties? Secondly, height is now fixed making the particles effect less responsive. How can I easily fix this issue?

View below:

<img width="109" alt="image" src="https://user-images.githubusercontent.com/79381943/144944617-746c3954-3090-4e23-bf4d-1088801e2296.png">

**Reply by matteobruni** [reply]
The parallax effect should work also in non-fullScreen container. About the height, once you remove the fullScreen feature, you have to handle the CSS by yourself, position, size, margin, and whatever you need.

**Reply by norvalbv** [reply]
Thanks. The parallax is not working for me. I'm either presuming because it's within a set sized container or because it's turned off within the options. The links on your README are outdated so I was wondering if you could link me to the options available?

**Reply by matteobruni** [reply]
This is a fixed size container with parallax: https://codepen.io/matteobruni/pen/NWaNNJG

The link in the readme is working: https://github.com/matteobruni/tsparticles#options, here there's an url pointing to the documentation here: https://particles.js.org/docs/interfaces/Options_Interfaces_IOptions.IOptions.html

**Reply by norvalbv** [reply]
Thank you for sending this over. Am still a little confused because it doesn't work as expected for me lol. Will have to read more of the docs.

In the meantime, when on fullscreen the particles cover components and some do not. Z-index doesn't come into play here nor background colours. Is there a reason for this? Each of the components are structured the same in terms of hierarchy

<img width="309" alt="image" src="https://user-images.githubusercontent.com/79381943/145038398-b0105e76-6131-467b-8fcb-76e0ea133980.png">

**Reply by matteobruni** [reply]
In the `fullScreen` options there's a `zIndex` property that defaults to `0` as you can see here: https://particles.js.org/docs/interfaces/Options_Interfaces_FullScreen_IFullScreen.IFullScreen.html

I think it's a `z-index` issue, I'd recommend `-1` as a value for using the particles as an animated background. They will be below everything and each transparent component will have them in background

**Comment by WilliamBotbyl** [comment]
how do I make the background stars instead of the default? I tried looking at other questions but can't figure it out

**Reply by matteobruni** [reply]
There’s the configuration link here: https://github.com/matteobruni/tsparticles/discussions/2378

**Comment by magicalunicor** [comment]
hello. I am using nextjs to integrate ts particles but I cannot get it working.
if I use [this "method"](https://dev.to/matteobruni/how-to-use-tsparticles-52k), as much as it looks like it would work because my background color changes, I see nothing. I have tried different configs but nothing.


if instead i write
`import { loadFull } from "tsparticles";`
it gives me:
`Syntax Error: cannot use import statement outside of a module`.

I have no idea about what to do. On React everything works correctly but on Next JS it seems like it doesn't. And of course, it's just me, because I've seen other projects with Next JS and they work correctly. I have no idea about what I'm doing wrong. But what I did is: install tsparticles using `npm install react-tsparticles` then I just paste this:
`import Particles from "react-tsparticles";`

`import particlesConfig from "../config/particle-config";`

`<Particles options={particlesConfig} />`


I don't know what to do. Please help me, I'm going crazy.

**Reply by magicalunicor** [reply]
edit: [this solution](https://stackoverflow.com/questions/66244968/cannot-use-import-statement-outside-a-module-error-when-importing-react-hook-m/66246141#66246141) worked for me. I had to install `next-transpile-modules` and add this two lines of code inside my next.config.js:
```
const withTM = require('next-transpile-modules')(['tsparticles']);

module.exports = withTM(nextConfig);
```

**Reply by matteobruni** [reply]
A fix is coming ASAP, this was reported in issue #3909

**Comment by jzfcoder** [comment]
I'm using tsparticles in react, and I can't see any other divs:
```
return (
  <>
    <Particles ... />
    <div>This is not visible</div>
  </>
);
```
even with fullScreen enable set to false, I can't see anything other than the Particles component; any ideas?

**Reply by matteobruni** [reply]
Can you share a reproducible sample? Disabling the fullScreen make particles act like any other div, so the CSS could be the issue

**Reply by jzfcoder** [reply]
Nevermind, I realized that I hadn't set position to absolute alongside zIndex. Thanks for help, apologies for bothering

**Comment by matteobruni** [comment]
This discussion will be closed since there's a dedicated repository for the React component: https://github.com/tsparticles/react
